security/src/main/scala/pl/touk/nussknacker/ui/security/api/AuthenticationManager.scala

```diff
@@ -0,0 +1,75 @@
+package pl.touk.nussknacker.ui.security.api
+
+import akka.http.scaladsl.server.Directive1
+import io.circe.parser._
+import io.circe.syntax._
+import pl.touk.nussknacker.security.AuthCredentials.{
+  ImpersonatedAuthCredentials,
+  NoCredentialsProvided,
+  PassedAuthCredentials
+}
+import pl.touk.nussknacker.security.{AuthCredentials, ImpersonatedUserData}
+import pl.touk.nussknacker.ui.security.accesslogic.{AnonymousAccess, ImpersonatedAccess}
+import sttp.tapir._
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class AuthenticationManager(authenticationResources: AuthenticationResources) {
+
+  lazy val authenticationRules: List[AuthenticationConfiguration.ConfigRule] =
+    authenticationResources.configuration.rules
+  private val anonymousAccess    = new AnonymousAccess(authenticationResources)
+  private val impersonatedAccess = new ImpersonatedAccess(authenticationResources)
+
+  def authenticate(): Directive1[AuthenticatedUser] = {
+    anonymousAccess.anonymousUserRole match {
+      case Some(role) =>
+        authenticationResources.authenticate().optional.flatMap {
+          case Some(user) => impersonatedAccess.provideOrImpersonateUser(user)
+          case None =>
+            anonymousAccess.handleAuthorizationFailedRejection.tmap(_ =>
+              AuthenticatedUser.createAnonymousUser(Set(role))
+            )
+        }
+      case None => authenticationResources.authenticate().flatMap(impersonatedAccess.provideOrImpersonateUser)
+    }
+  }
+
+  def authenticate(
+      authCredentials: AuthCredentials
+  )(implicit ec: ExecutionContext): Future[Option[AuthenticatedUser]] = {
+    authCredentials match {
+      case passedCredentials @ PassedAuthCredentials(_) => authenticationResources.authenticate(passedCredentials)
+      case NoCredentialsProvided                        => anonymousAccess.authenticateWithAnonymousAccess()
+      case ImpersonatedAuthCredentials(impersonatingUserAuthCredentials, impersonatedUserData) =>
+        impersonatedAccess.authenticateWithImpersonation(impersonatingUserAuthCredentials, impersonatedUserData)
+    }
+  }
+
+  def authenticationEndpointInput(): EndpointInput[AuthCredentials] = {
+    authenticationResources
+      .authenticationMethod()
+      .and(ImpersonatedAccess.headerEndpointInput)
+      .map(
+        Mapping.fromDecode[(Option[String], Option[String]), AuthCredentials] {
+          case (Some(passedCredentials), None) => DecodeResult.Value(PassedAuthCredentials(passedCredentials))
+          case (Some(passedCredentials), Some(impersonatedUserData)) =>
+            decode[ImpersonatedUserData](impersonatedUserData) match {
```

Comment:
I'm not sure if we should pass all the impersonated user data. I suspect that this is done for the sake of oauth2, right? 

In the case of basic auth it's simple because we have the user on the backend side. We can (and TBH we should) verify if the user exists and has proper permissions. For oauth2 it's not a case.

Reply:
As I mentioned in one of the comments https://github.com/TouK/nussknacker/pull/6053#discussion_r1622667524 new mechanism with `ImpersonationContext` was introduced. Impersonation is currently implemeneted for Basic Auth and I left a TODO comment for OAuth2.